File: Top_100_Liked_Questions/215._Kth_Largest_Element_in_an_Array/solution.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var findKthLargest = function(nums, k) {
    // quick select
    let partition = function(left, right, pivot_idx) {
        let pivot_val = nums[pivot_idx];
        [ nums[right], nums[pivot_idx] ] = [ nums[pivot_idx], nums[right] ];
        let new_pivot_idx = left;
        for (let i=left; i<right; i++) {
            if (nums[i] < pivot_val) {
                [ nums[i], nums[new_pivot_idx] ] = [ nums[new_pivot_idx], nums[i] ];
                new_pivot_idx++;
            }
        }
        [ nums[new_pivot_idx], nums[right] ] = [ nums[right], nums[new_pivot_idx] ];
        return new_pivot_idx;
    }
    
    let select = function(left, right, k_smallest) {
        if (left === right) {
            return nums[left];
        }
        
        let pivot_idx = left + Math.floor(Math.random() * (right - left + 1));
        pivot_idx = partition(left, right, pivot_idx);
        
        if (pivot_idx === k_smallest) {
            return nums[pivot_idx];
        } else if (k_smallest < pivot_idx) {
            return select(left, pivot_idx-1, k_smallest);
        } else {
            return select(pivot_idx+1, right, k_smallest);
        }
    }
    
    return select(0, nums.length-1, nums.length-k);
};

var findKthLargest3 = function(nums, k) {
    // min heap
    class minHeap {
        static heapify(arr) {
            for (let i=arr.length; i>-1; i--) {
                minHeap.siftDown(arr, i);
            }
            return arr;
        }
        
        static insert(heap, num) {
            heap.push(num);
            minHeap.siftUp(heap, heap.length-1);
        }
        
        static pop(heap) {
            let top = heap[0];
            heap[0] = heap.pop();
            minHeap.siftDown(heap, 0);
            return top;
        }
        
        static siftUp(heap, idx) {
            while (idx > 0) {
                let par_idx = minHeap.getParentIdx(heap, idx);
                if (heap[idx] < heap[par_idx]) {
                    [ heap[idx], heap[par_idx] ] = [ heap[par_idx], heap[idx] ];
                    idx = par_idx;
                } else {
                    break;
                }
            }
        }
        
        static siftDown(heap, idx) {
            while (idx < heap.length) {
                let child_idx = minHeap.getMinChildIdx(heap, idx);
                if (!child_idx) {
                    break;
                }
                
                if (heap[idx] > heap[child_idx]) {
                    [ heap[idx], heap[child_idx] ] = [ heap[child_idx], heap[idx] ];
                    idx = child_idx;
                } else {
                    break;
                }
            }
        }
        
        static getParentIdx(idx) {
            return Math.floor((idx-1)/2);
        }
        
        static getMinChildIdx(heap, idx) {
            let left = idx * 2 + 1;
            let right = idx * 2 + 2;
            
            if (left >= heap.length) {
                return null;
            }
            
            if (right >= heap.length) {
                return left;
            }
            
            return heap[left] < heap[right] ? left : right;
        }
    }
    
    let heap = minHeap.heapify(nums.slice(0,k));
    
    for (let i=k; i<nums.length; i++) {
        if (nums[i] > heap[0]) {
            minHeap.insert(heap, nums[i]);
            minHeap.pop(heap);
        }
    }
    
    return heap[0];
};

var findKthLargest2 = function(nums, k) {
    // sorting
    return nums.sort((a,b) => b - a)[k-1];
};

File: Top_100_Liked_Questions/215._Kth_Largest_Element_in_an_Array/solution.py
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        # quick select - O(n) average, O(n**2) worst
        def partition(left, right, piv_idx):
            piv_val = nums[piv_idx]
            nums[piv_idx], nums[right] = nums[right], nums[piv_idx]
            st_idx = left
            for i in range(left, right):
                if nums[i] < piv_val:
                    nums[st_idx], nums[i] = nums[i], nums[st_idx]
                    st_idx += 1
            
            nums[right], nums[st_idx] = nums[st_idx], nums[right]
            return st_idx
        
        
        def select(left, right, k_smallest):
            if left == right:
                return nums[left]
            
            pivot_index = random.randint(left, right)
            pivot_index = partition(left, right, pivot_index)
            
            if k_smallest == pivot_index:
                 return nums[k_smallest]
            elif k_smallest < pivot_index:
                return select(left, pivot_index - 1, k_smallest)
            else:
                return select(pivot_index + 1, right, k_smallest)
 
        return select(0, len(nums) - 1, len(nums) - k)
        
    
    def findKthLargest3(self, nums: List[int], k: int) -> int:
        # minHeap - O(k + (n - k)log(k))
        return heapq.nlargest(k, nums)[-1]
    
    
    def findKthLargest2(self, nums: List[int], k: int) -> int:
        # sorting - O(n*log(n))
        return sorted(nums, reverse=True)[k-1]